lib/config.js

```diff
@@ -1,14 +1,43 @@
+'use strict';
+
 var config = {
-    brand: 'Discord',
-    token: process.env.OAUTH_TOKEN,
-    trackingID: process.env.TRACKING_ID,
-    commentWait: process.env.COMMENT_WAIT || 250,
-    commentAttempts: process.env.COMMENT_ATTEMPTS || 10,
-
-    // Used when no environment variable (process.env.PORT) is set
-    port: 3000,
-    host: 'localhost',
-    protocol: 'http:'
+    def: {
+        brand: 'Discord',
+        token: process.env.OAUTH_TOKEN,
+        commentAttempts: process.env.COMMENT_ATTEMPTS || 10,
+    },
+    development: {
+        port: 3000,
+        redisURL: 'redis://localhost',
+        databaseURL: 'postgres://localhost/discord_test'
+    },
+    test: {
+        protocol: 'http:',
+        host: 'localhost',
+        port: 3000,
+        trackingID: '654321',
+        commentWait: 0,
+        redisURL: 'redis://localhost',
+        databaseURL: 'postgres://localhost/discord_test',
+        database: 'discord_test'
+    },
+    production: {
+        port: process.env.PORT,
+        trackingID: process.env.TRACKING_ID,
+        commentWait: process.env.COMMENT_WAIT || 250,
+        redisURL: process.env.REDIS_URL,
+        databaseURL: process.env.DATABASE_URL
+    }
```

Comment:
Thanks for mentioning this. I misread [12factor.net](http://12factor.net/) as recommending that only sensitive data (like database passwords) be stored in environment variables. I also noticed that Sequelize [uses a configuration file just like this](https://gist.github.com/openjck/bdbe892ee2ee44444e25) and tried to follow that pattern. But that's a case of me conflating Node and 12factor.

`autoenv` sounds like a better choice, so I'm happy to use that instead. At the same time, I was curious about something and was wondering if you could say more. I'm having trouble understanding how a *.env* file and a *config.json* really differ at the end of the day. I don't doubt that `autoenv` is the way to go, but I'm having trouble understanding why from a 12factor point of view.

On a more practical level, how would precedence work with `autoenv`? I'm having trouble understanding how we would set local settings that wouldn't also apply on production.